cockatrice/src/handle_public_servers.cpp

```diff
@@ -0,0 +1,92 @@
+//
+// Created by Zachary Halpern on 4/6/18.
+//
+
+#include "handle_public_servers.h"
+#include "qt-json/json.h"
+#include "settingscache.h"
+#include <QUrl>
+
+#define PUBLIC_SERVERS_JSON "https://cockatrice.github.io/public-servers.json"
+
+HandlePublicServers::HandlePublicServers(QObject *parent) : QObject(parent), nam(nullptr), reply(nullptr)
+{
+}
+
+void HandlePublicServers::downloadPublicServers()
+{
+    QUrl url(QString(PUBLIC_SERVERS_JSON));
+    nam = new QNetworkAccessManager(this);
+    reply = nam->get(QNetworkRequest(url));
+    connect(reply, SIGNAL(finished()), this, SLOT(actFinishParsingDownloadedData()));
+}
+
+void HandlePublicServers::actFinishParsingDownloadedData()
+{
+    reply = dynamic_cast<QNetworkReply *>(sender());
+    QNetworkReply::NetworkError errorCode = reply->error();
+
+    // After finished() is called, this object will be deleted
+    nam->deleteLater();
+
+    if (errorCode == QNetworkReply::NoError) {
+        // Get current saved hosts
+        UserConnection_Information uci;
+        savedHostList = uci.getServerInfo();
+
+        // List of public servers
+        // Will be populated below
+        QStringList publicServersAvailable;
+        QStringList publicServersToRemove;
+
+        // Downloaded data from GitHub
+        QString jsonData = QString(reply->readAll());
+        auto jsonMap = QtJson::Json::parse(jsonData).toMap();
+
+        // Servers available
+        auto publicServersJSONList = jsonMap["servers"].toList();
+        for (const auto &server : publicServersJSONList) {
+            // Data inside one server at a time
+            // server: [{ ... }, ..., { ... }]
+            const auto serverMap = server.toMap();
+
+            QString serverName = serverMap["name"].toString();
+
+            if (serverMap["isInactive"].toBool()) {
+                publicServersToRemove.append(serverName);
+                continue;
+            }
+
+            QString serverAddress = serverMap["host"].toString();
+            QString serverPort = serverMap["port"].toString();
+
+            publicServersAvailable.append(serverName);
+
+            if (!savedHostList.contains(serverName)) {
```

Comment:
What about checking for server url instead of name? That seems more likely to be accurate.

Reply:
The host list is a QList<QString> of server names only... I'd rather not change that around if I can avoid it

Reply:
std::pair it up?